modules/planning/tasks/deciders/path_bounds_decider.cc

```diff
@@ -16,19 +16,174 @@
 
 #include "modules/planning/tasks/deciders/path_bounds_decider.h"
 
+#include <algorithm>
+#include <limits>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "modules/common/configs/vehicle_config_helper.h"
+
 namespace apollo {
 namespace planning {
 
+using apollo::common::ErrorCode;
 using apollo::common::Status;
+using apollo::common::VehicleConfigHelper;
+
+constexpr double kPathBoundsDeciderHorizon = 100.0;
+constexpr double kPathBoundsDeciderResolution = 0.1;
+constexpr double kDefaultLaneWidth = 5.0;
+constexpr double kDefaultRoadWidth = 20.0;
+constexpr double kRoadEdgeBuffer = 0.2;
 
 PathBoundsDecider::PathBoundsDecider(const TaskConfig &config)
     : Decider(config) {}
 
 Status PathBoundsDecider::Process(
     Frame *const frame, ReferenceLineInfo *const reference_line_info) {
-  // TODO(all) implement
+  // Sanity checks.
+  CHECK_NOTNULL(frame);
+  CHECK_NOTNULL(reference_line_info);
+
+  // The decided path bounds should be in the format of: (s, l_min, l_max).
+  std::vector<std::tuple<double, double, double>> path_boundaries;
+
+  // 0. Initialize the path boundaries to be an indefinitely large area.
+  if (!InitPathBoundaries(
+          reference_line_info->reference_line(),
+          frame->PlanningStartPoint(), &path_boundaries)) {
+    const std::string msg = "Failed to initialize path boundaries.";
+    AERROR << msg;
+    return Status(ErrorCode::PLANNING_ERROR, msg);
+  }
+  // 1. Decide a rough boundary based on road info and ADC's position
+  if (!GetBoundariesFromRoadsAndADC(
+          reference_line_info->reference_line(),
+          reference_line_info->AdcSlBoundary(), &path_boundaries)) {
+    const std::string msg = "Failed to decide a rough boundary based on "
+                            "road information.";
+    AERROR << msg;
+    return Status(ErrorCode::PLANNING_ERROR, msg);
+  }
+  // 2. Finetune the boundary based on static obstacles
+  // TODO(all): in the future, add side-pass functionality.
+  if (!GetBoundariesFromStaticObstacles(
+          reference_line_info->path_decision()->obstacles(),
+          &path_boundaries)) {
+    const std::string msg = "Failed to decide fine tune the boundaries after "
+                            "taking into consideration all static obstacles.";
+    AERROR << msg;
+    return Status(ErrorCode::PLANNING_ERROR, msg);
+  }
+
+  // Update the path boundary info to the frame.
+  // TODO(all): update frame with path bound info.
   return Status::OK();
 }
 
+bool PathBoundsDecider::InitPathBoundaries(
+    const ReferenceLine& reference_line,
+    const common::TrajectoryPoint& planning_start_point,
+    std::vector<std::tuple<double, double, double>>* const path_boundaries) {
+  // Sanity checks.
+  CHECK_NOTNULL(path_boundaries);
```

Comment:
In which case this function will return false? If not, it is cleaner to return the path_boundaries directly other than passing it as a parameter.

Reply:
Same reason as below.